File: 972-knight-dialer/knight-dialer.cpp
class Solution {
public:
    int MOD_VAL = 1e9 + 7;
    vector<vector<int>> moves = {
        {4, 6},
        {6, 8},
        {7, 9},
        {4, 8},
        {0, 3, 9},
        {},
        {1, 7, 0},
        {2, 6},
        {1, 3},
        {2, 4}
    };

    int backtrack(int n, int cur, vector<vector<int>>& memo) {
        if(n == 0) return 1;
        if(memo[n][cur] != -1) return memo[n][cur];

        int ways = 0;
        for(auto ele : moves[cur]) {
            ways = ((ways + backtrack(n - 1, ele, memo)) % MOD_VAL);
        }
        return memo[n][cur] = ways;
    }

    int knightDialer(int n) {
        vector<vector<int>> memo(n + 1, vector<int>(10, -1));
        if(n == 0) return 0;
        
        int ans = 0;
        for(int i = 0; i < 10; i++) {
            ans = ((ans + backtrack(n - 1, i, memo)) % MOD_VAL);
        }
        return ans;
    }
};

File: 972-knight-dialer/knight-dialer.py
MOD_VAL = 10 ** 9 + 7
validMoves = {
    1: [6, 8],
    2: [7, 9],
    3: [4, 8],
    4: [0, 3, 9],
    5: [],
    6: [7, 0, 1],
    7: [2, 6],
    8: [1, 3],
    9: [2, 4],
    0: [4, 6],
}

class Solution:            
    def knightDialer(self, n: int) -> int:
        cache = {}
        def backtrack(n, cur):
            if n == 0:
                return 1
            
            if (n, cur) in cache:
                return cache[(n, cur)]
            
            ways = 0
            for nextOption in validMoves[cur]:
                ways = (ways + backtrack(n - 1, nextOption)) % MOD_VAL
            cache[(n, cur)] = ways
            return ways

        if n == 0:
            return 0
        
        ans = 0
        for i in range(10):
            ans = (ans + backtrack(n - 1, i)) % MOD_VAL
        return ans  